node/src/components/storage/lmdb_ext.rs

```diff
@@ -0,0 +1,157 @@
+}
+
+// Classifies an `lmdb::Error` according to our scheme. This one of the rare cases where we accept a
+// blanked `From<>` implementation for error type conversion.
+impl From<lmdb::Error> for LmdbExtError {
+    fn from(lmdb_error: lmdb::Error) -> Self {
+        match lmdb_error {
+            lmdb::Error::PageNotFound
+            | lmdb::Error::Corrupted
+            | lmdb::Error::Panic
+            | lmdb::Error::VersionMismatch
+            | lmdb::Error::Invalid
+            | lmdb::Error::Incompatible => LmdbExtError::LmdbCorrupted(lmdb_error),
+
+            lmdb::Error::MapFull
+            | lmdb::Error::DbsFull
+            | lmdb::Error::ReadersFull
+            | lmdb::Error::TlsFull
+            | lmdb::Error::TxnFull
+            | lmdb::Error::CursorFull
+            | lmdb::Error::PageFull
+            | lmdb::Error::MapResized => LmdbExtError::ResourceExhausted(lmdb_error),
+
+            lmdb::Error::NotFound
+            | lmdb::Error::BadRslot
+            | lmdb::Error::BadTxn
+            | lmdb::Error::BadValSize
+            | lmdb::Error::BadDbi
+            | lmdb::Error::KeyExist
+            | lmdb::Error::Other(_) => LmdbExtError::Other(Box::new(lmdb_error)),
+        }
+    }
+}
+
+/// Additional methods on transaction.
+pub(super) trait TransactionExt {
+    /// Helper function to load a value from a database.
+    fn get_value<K: AsRef<[u8]>, V: DeserializeOwned>(
+        &mut self,
+        db: Database,
+        key: &K,
+    ) -> Result<Option<V>, LmdbExtError>;
+}
+
+/// Additional methods on write transactions.
+pub(super) trait WriteTransactionExt {
+    /// Helper function to write a value to a database.
+    ///
+    /// Returns `true` if the value has actually been written, `false` if the key already existed.
+    ///
+    /// Setting `overwrite` to true will cause the value to always be written instead.
+    ///
+    /// # Panics
+    ///
+    /// Panics if a database error occurs.
+    fn put_value<K: AsRef<[u8]>, V: Serialize>(
+        &mut self,
+        db: Database,
+        key: &K,
+        value: &V,
+        overwrite: bool,
+    ) -> Result<bool, LmdbExtError>;
+}
+
+impl<T> TransactionExt for T
+where
+    T: Transaction,
+{
+    #[inline]
+    fn get_value<K: AsRef<[u8]>, V: DeserializeOwned>(
+        &mut self,
+        db: Database,
+        key: &K,
+    ) -> Result<Option<V>, LmdbExtError> {
+        match self.get(db, key) {
+            // Deserialization failures are likely due to storage corruption.
+            Ok(raw) => deser(raw).map(Some),
+            Err(lmdb::Error::NotFound) => Ok(None),
+            Err(err) => Err(err.into()),
+        }
+    }
+}
+
+impl WriteTransactionExt for RwTransaction<'_> {
+    fn put_value<K: AsRef<[u8]>, V: Serialize>(
+        &mut self,
+        db: Database,
+        key: &K,
+        value: &V,
+        overwrite: bool,
+    ) -> Result<bool, LmdbExtError> {
+        let buffer = ser(value).expect("TODO");
```

Comment:
Apparently my comment on this line got missed from my review.

Anyway - this line also violates the no panic rule, but if the `expect` is to be retained, I reckon you could come up with a more helpful error message :smile:

Reply:
Already turned it into an error, see above :)